server/queues/processors/NotificationsProcessor.ts

```diff
@@ -182,3 +225,85 @@ export default class NotificationsProcessor extends BaseProcessor {
     }
   }
 }
+
+type Notifiable = {
+  user: User;
+  userId: string;
+  document: Document;
+  documentId: string;
+  event: DocumentActionEvent | RevisionEvent;
+  eventName: string;
+};
+
+const shouldNotify = async (subject: Notifiable): Promise<boolean> => {
+  // Suppress notifications for suspended users
+  if (subject.user.isSuspended) {
+    return false;
+  }
+
+  // Check the user has access to the collection this document is in.
+  // Just because they were a collaborator once doesn't mean they still are.
+  const collectionIds = await subject.user.collectionIds();
+  if (!collectionIds.includes(subject.document.collectionId)) {
+    return false;
+  }
+
+  // If this user has viewed the document since the last update was made
+  // then we can avoid sending them a useless notification, yay.
+  const view = await View.findOne({
+    where: {
+      userId: subject.userId,
+      documentId: subject.event.documentId,
+      updatedAt: {
+        [Op.gt]: subject.document.updatedAt,
+      },
+    },
+  });
+
+  if (view) {
+    Logger.info(
+      "processor",
+      `suppressing notification to ${subject.userId} because update viewed`
+    );
+    return false;
+  }
+
+  if (!subject.user.email) {
+    return false;
+  }
+
+  return true;
+};
+
+const createSubscriptions = async (
+  document: Document,
+  event: DocumentEvent
+): Promise<void> => {
+  const collaboratorIds = document.collaboratorIds;
+
+  for (const collaboratorId of collaboratorIds) {
```

Comment:
Can parallelize this with `Promise.all` and speed up the processor significantly